lang/strings.ts

```diff
@@ -0,0 +1,86 @@
+import { EOL, isNil } from '../deps.ts';
+
+/**
+ * Removes a *single* trailing new line (either LF or CRLF) from the string, if there is one.
+ *
+ * Note: Unlike the `string.trimEnd()` function, this will not remove any other whitespace
+ * characters.
+ *
+ * @param text The text to be altered
+ * @returns The altered text
+ */
+export function trimTrailingNewline(text: string): string {
+  if (text.endsWith(EOL.CRLF)) {
+    return stripSuffix(text, EOL.CRLF);
+  } else {
+    return stripSuffix(text, EOL.LF);
+  }
+}
+
+/**
+ * Removes the specified prefix from the string if it exists.
+ *
+ * @param text The text to be altered
+ * @param prefix The prefix to remove from the text
+ * @returns The altered text
+ */
+export function stripPrefix(text: string, prefix: string): string {
+  return text.startsWith(prefix) ? text.substring(prefix.length) : text;
+}
+
+/**
+ * Removes the specified suffix from the string if it exists.
+ *
+ * @param text The text to be altered
+ * @param suffix The suffix to remove from the text
+ * @returns The altered text
+ */
+export function stripSuffix(text: string, suffix: string): string {
+  return text.endsWith(suffix) ? text.substring(0, text.length - suffix.length) : text;
+}
+
+/**
+ * Tag template function that formats the template string with the leading/trailing newlines and
+ * largest common line indentations removed.
+ *
+ * @param literals The literals in the template string (provided by tagged template)
+ * @param values The values in the template string (provided by tagged template)
+ * @returns The rendered and formatted string
+ *
+ * @example
+ * const generatedCode = trimAndDedent`
+ *   service ${serviceName} {
+ *     rpc Delete${subjectName}(Delete${subjectName}Request) returns (Delete${subjectName}Response);
+ *     rpc Create${subjectName}(Create${subjectName}Request) returns (Create${subjectName}Response);
+ *   }
+ * `;
+ */
+export function trimAndDedent(literals: TemplateStringsArray, ...values: string[]): string {
```

Comment:
This is maybe a deno-ism or javascript-ism that I'm unfamiliar with, but why make this a tag template function instead of a regular function, especially because it's not doing anything particularly special requiring the tag template functionality?

Reply:
Yeah, it definitely doesn't do anything useful with the template tags. The idea of making it a template tag felt right to me because it only makes sense to use it on multiline string literals, which would always be template strings in JS. While I feel like I've seen this done a lot before, I guess that doesn't feel like a great reason. Converted to a regular function.